Gp/size estimation z operations

JackPiri opened the pull request:
Pull request associated to issue https://github.com/HorizenOfficial/zen/issues/475

JackPiri committed:
WalletMergeToAddressTest test enhancement

More detailed test for different scenarios (t2t, t2z, z2t, z2z, tz2t, tz2z)

JackPiri committed:
FIx for z_mergetoaddress transaction size estimation

Now the estimated number of joinsplits is aligned with the number of the joinsplit actually contained in transaction
Extended Python tests for z_mergetoaddress

AndreaLanfranchi commented on src/wallet/rpcwallet.cpp:
```diff
@@ -5946,7 +5946,7 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
 #define MERGE_TO_ADDRESS_DEFAULT_SHIELDED_LIMIT 10
 
-#define JOINSPLIT_SIZE JSDescription::getNewInstance(GROTH_TX_VERSION).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION, GROTH_TX_VERSION)
+#define JOINSPLIT_SIZE(TX_VER) JSDescription::getNewInstance(TX_VER == GROTH_TX_VERSION).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION, TX_VER)
```
Macros like this should be avoided. Let's take the opportunity to remove them and transform in a static function

JackPiri replied:
Done

a-petrini commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,348 @@
+        with zipfile.ZipFile(resource_file, 'r') as zip_ref:
+            zip_ref.extractall(self.options.tmpdir)
+        copy_tree(self.options.tmpdir + "/test_setup_", self.options.tmpdir)
+        shutil.rmtree(self.options.tmpdir + "/test_setup_")
```
`import os` and `import zipfile` can be moved as global import. Also, be removing the `test_setup_` folder as root of the zip file (i.e. the root of the zip will contain the directories `node0`, `node1` and `node2`) it is possible to use the `extractall()` method without the need of do `chdir()`, `copy_tree()` and `rmtree()`.
Finally, using the `join()` function with `os.sep()` as join element will result in a more portable code
```suggestion
    def import_data_to_data_dir(self):
        #importing datadir resource
        resource_file = os.sep.join([os.path.dirname(__file__), 'resources', 'wallet_mergetoaddress_2', 'test_setup_.zip'])
        with zipfile.ZipFile(resource_file, 'r') as zip_ref:
            zip_ref.extractall(self.options.tmpdir)
```

JackPiri replied:
Done

a-petrini commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,348 @@
+    wait_and_assert_operationid_status
+
+import time
+from decimal import Decimal
```
There are a few unused imports that can be removed, in particular `JSONRPCException`, `sync_blocks, sync_mempools`, `time` and `Decimal`

JackPiri replied:
Done

a-petrini commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,348 @@
+        for i in range(quantity):
+            self.nodes[1].sendmany("", {self.nodes[0].listaddresses()[0]: 10}, 100)
+        self.nodes[1].generate(100)
+        self.sync_all
```
This is a bug, as it is not invoking the sync_all function
```suggestion
        self.sync_all()
```

JackPiri replied:
Done

a-petrini commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,348 @@
+                    break
+                if (bef == len(unspent_transactions_before_merge) - 1):
+                    txid = unspent_transactions_after_merge[aft]["txid"]
+        return self.nodes[2].gettransaction(txid)
```
This can be simplified using the `in` keyword to check if a specific element is present in a list or dictionary, given that the input lists differ for at least one element. The suggested function returns `None` if the lists contain the same elements.
```suggestion
    def get_merged_transaction(self, unspent_transactions_before_merge, unspent_transactions_after_merge):
        for tx in unspent_transactions_after_merge:
            if tx not in unspent_transactions_before_merge:
                return self.nodes[2].gettransaction(tx["txid"])
        return None
```

a-petrini replied:
As a follow up, the previous suggested change compares the txs in both lists. If you need to compare only the "txid" of each transaction, the function should be this one:
```suggestion
    def get_merged_transaction(self, unspent_transactions_before_merge, unspent_transactions_after_merge):
        unspent_txid_before = [x["txid"] for x in unspent_transactions_before_merge]
        for txid in [x["txid"] for x in unspent_transactions_after_merge]:
            if txid not in unspent_txid_before:
                return self.nodes[2].gettransaction(txid)
        return None
```

JackPiri replied:
Done

a-petrini commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,348 @@
+        balance_before_merge = self.nodes[2].z_getbalance(mytaddr2)
+        result = self.nodes[0].z_mergetoaddress(["ANY_TADDR"], mytaddr2, 0, utxos_quantity, 0)        
+        wait_and_assert_operationid_status(self.nodes[0], result['opid'], "success", "", 3600)
+        self.nodes[1].generate(100)
```
I don't know if mining 100 blocks is really a necessity. If it is done to prevent some timing issues, maybe it is possible to main blocks (i.e. mine a block and check the condition) until the issue is solved.

JackPiri replied:
Moved to single block mining on node1 (apart from `generate_notes` method).

AndreaLanfranchi reviewed (approved):
LGTM

JackPiri committed:
wallet_mergetoaddress_2.py test fix

Now it is possible to set explicitly the maximum tip age, allowing the usage of old datadir setup

JackPiri committed:
Possibility to set single test execution on Travis

Useful for launching separately long execution tests

drgora commented on src/chainparams.cpp:
```diff
@@ -394,6 +394,15 @@ bool SelectParamsFromCommandLine()
+    if ( Params().NetworkIDString() == "regtest" )
+    {
+        int maxTipAge = (int)GetArg("-maxtipage", nMaxTipAge);
+        return maxTipAge;
```
Any reason for the cast to `int` and then to `int64_t`? It looks like you can make it more compact as:

```suggestion
        return GetArg("-maxtipage", nMaxTipAge);
```
Also, please try to keep a style consistent with the rest of the file.

JackPiri replied:
Done

drgora commented on src/wallet/rpcwallet.cpp:
```diff
@@ -5946,7 +5946,9 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
-#define JOINSPLIT_SIZE JSDescription::getNewInstance(GROTH_TX_VERSION).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION, GROTH_TX_VERSION)
+int GetJoinSplitSize(int shieldedTxVersion) {
+    return JSDescription::getNewInstance(shieldedTxVersion == GROTH_TX_VERSION).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION, shieldedTxVersion);
+}
```
Just a small thing here; I'd suggest to put this function in an anonymous namespace (or alternatively make it `static`), so that its visibility remains limited to this file, given that it's not used elsewhere.

```suggestion
namespace {

int GetJoinSplitSize(int shieldedTxVersion) {
    return JSDescription::getNewInstance(shieldedTxVersion == GROTH_TX_VERSION).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION, shieldedTxVersion);
}

}
```

JackPiri replied:
Done

ptagl commented on qa/pull-tester/rpc-tests.sh:
```diff
@@ -16,6 +16,10 @@ for i in "$@"; do
       EXTENDED="true"
       shift
       ;;
+    -specific=*)
```
I thinks it's great now to have the flexibility to run one specific test now, I just wonder if we can rename the `-specific` keyword to something else to improve clarity.

I don't have any particular keyword in mind, so I would like to have feedback from other people.

JackPiri replied:
Maybe something like `-runsingletest` can be more clear?

ptagl replied:
As a temporary proposal we could use `-runonly`. We could also consider to refactor the script to have a regex parameter to filter the list of tests.

JackPiri replied:
Done

ptagl commented on qa/zcash/full_test_suite.py:
```diff
@@ -192,6 +192,8 @@ def main():
     parser.add_argument('--rpc-split', dest='split',
                         action='store', help='string in format m:n, see qa/rpc-tests/README.md for more')
+    parser.add_argument('--rpc-specific', dest='specific',
+                        action='store', help='execute only a specific python test, see qa/rpc-tests/README.md for more')                        
```
Shall we add information to the README file about the usage of this argument (eventually after its renaming)?

ptagl commented on src/init.cpp:
```diff
@@ -609,6 +609,9 @@ std::string HelpMessage(HelpMessageMode mode)
     }
 
+    strUsage += HelpMessageOpt("-maxtipage=<n>",
+        "regtest only - Set the maximum tip age)");
```
```suggestion
        "regtest only - Set the maximum tip age used for the relay of chain blocks)");
```

JackPiri replied:
Done

ptagl commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,341 @@
+        assert_equal(result[self.op_result_k][self.mergingShieldedValue_k], 0)
+        #no joinsplit is required because every input is transparent and every output is transparent
+        assert_equal(result[self.joinsplit_len_k], 0)
+        # ---------- T to T ----------
```
Is there any specific reason to test `z_mergetoaddress_t2t(3)` and then `z_mergetoaddress_t2t(4)` immediately after? Do they represent special cases?

JackPiri replied:
Removed all tests apart from oversized transaction test

ptagl commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,341 @@
+    mergingShieldedValue_k = "mergingShieldedValue"
+
+    def import_data_to_data_dir(self):
+        #importing datadir resource
```
Can we add a white space at the beginning of any comment to improve readability?

JackPiri replied:
Done

ptagl commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,341 @@
+        self.sync_all()
+
+    def run_test (self):
+
```
What about extending the current description to explain why this Python test has been developed? If I remember correctly, it was to test specifically a change to the `z_mergetoaddress()` RPC command that fixed the estimation of the final transaction size. Is it right?

JackPiri replied:
Yes, added desciption of the test at the start of `run_test` (both for `wallet_mergetoaddress.py` as for `wallet_mergetoaddress`)

ptagl commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,341 @@
+        #57 joinsplits are required because first two shielded inputs are merged in a shielded output,
+        #which is then set as shielded input (change) in the subsequent joinsplit and paired with another shielded input
+        #until the final shielded output is obtained
+        assert_equal(result[self.joinsplit_len_k], 57)
```
I guess that this test is important to check the behavior of the implemented fix. I would expect that this test was previously failing due to a wrong estimation of the transaction size, while after the fix works as expected. Am I right?

If so, I'm not sure if all the previous tests and the last one are relevant to this PR.

JackPiri replied:
Removed other tests, moved tz2t test to `wallet_mergetoaddress.py`

drgora commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,341 @@
+    mergingTransparentValue_k = "mergingTransparentValue"
+    mergingShieldedValue_k = "mergingShieldedValue"
+
+    def import_data_to_data_dir(self):
```
I think it would be nice to add, either here or to the `run_test`, a description of the state being imported, e.g. in terms of chains height, num of utxos for each node and so...

Also, using a more specific name, compared to `test_setup_.zip`, might be more future proof in case we want to implement a similar approach in other tests.

JackPiri replied:
Done

drgora commented on qa/rpc-tests/wallet_mergetoaddress_2.py:
```diff
@@ -0,0 +1,341 @@
+        # Node1 acts as miner and funds distributor
+        # Node2 acts ad "to" (merge_to_address receiver)
+
+        #(generate 1b on n0, generate 100bs on node1, shield 1 b on n0) * 100 + generate 1b on n1
```
Does this line refer to the state imported with `import_data_to_data_dir`? If so, could you please make it explicit?

JackPiri replied:
Yes, it is related to the state imported with `import_data_to_data_dir`, added explicit comment on that function.

JackPiri committed:
Modifications after review

General refactoring

JackPiri committed:
Data dir setup (wallet_mergetoaddress_2.py)

Data dir is downloaded from URL instead of being archived within the repo
